### PR TITLE
feat(VDatePicker): native events on day/month/year

### LINKED ISSUE #4884: [Feature Request] Popup events on mouseover for VDatePicker
### What problem does this feature solve?
Users should be able to get a top level description of the events for the mouseover-ed date.

### What is your proposed solution?
Supply additional event @mouseover with the hovered date so the parent could display the details for the date.

Component: VDatePicker

### LINKED ISSUE #5598: [Feature Request] Expose hovered date value on VDatePicker
### Problem to solve
On the path to making a date-range picker for my app, similar to many booking type systems out there, I'd really benefit from having the VDatePicker expose what date is currently being hovered over (example reference: https://www.booking.com/).

This opens the way to implementing the rest of the features for such a date-range picker.

### Proposed solution
Implementing an event emitter to the VDatePicker, called @hover or something similar, that returns the current date being hovered over.

### PR DESCRIPTION
**Pull request by jacekkarczmarczyk**
## Motivation and Context
fixes #5598
fixes #4884

## How Has This Been Tested?
playground, unit

## Markup:

<details>

```vue
<template>
  <v-date-picker
    @contextmenu:month="(value, event) => e('contextmenu (with prevent default)', value, event)"
    @mouseenter:year="e('mouseenter', $event)"
    @dblclick:date="e('dblclick', $event)"
  />
</template>

<script>
  export default {
    methods: {
      e (name, value, event) {
        event && event.preventDefault() // for context menu example
        console.log(name, value)
      }
    }
  }
</script>
```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and backwards compatible changes and `next` for non-backwards compatible changes).
- [x] My code follows the code style of this project.
- [ ] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

## TODO
- [x] extract `reduce()` to utils

### CODE REVIEW

File: packages/vuetify/src/components/VDatePicker/VDatePicker.ts [hunk 1]
```diff
@@ -401,8 +401,12 @@ export default mixins(
         on: {
           input: this.dateClick,
           'update:table-date': (value: string) => this.tableDate = value,
-          'click:date': (value: string) => this.$emit('click:date', value),
-          'dblclick:date': (value: string) => this.$emit('dblclick:date', value),
+          ...Object.keys(this.$listeners).reduce((on, eventName) => {
+            if (eventName.match(/:date$/)) {
```
Comment (KaelWD) [inline packages/vuetify/src/components/VDatePicker/VDatePicker.ts:405]: `.endsWith`?

File: packages/vuetify/src/components/VDatePicker/VDatePicker.ts [hunk 2]
```diff
@@ -401,8 +401,12 @@ export default mixins(
...
+              Object.assign(on, { [eventName]: (value: string) => this.$emit(eventName, value) })
```
Comment (KaelWD) [inline packages/vuetify/src/components/VDatePicker/VDatePicker.ts:406]: `on[eventName] = ...`

Comment (jacekkarczmarczyk) [reply packages/vuetify/src/components/VDatePicker/VDatePicker.ts:406]: yup, i think i had immediate returning updated object in my mind

File: packages/docs/src/examples/date-pickers/simple/date-button-events.vue [hunk 3]
```diff
@@ -0,0 +1,47 @@
+<template>
```
Comment (johnleider) [inline packages/docs/src/examples/date-pickers/simple/date-button-events.vue:1]: Please improve the markup formatting.

File: packages/docs/src/lang/en/components/DatePickers.json [hunk 4]
```diff
@@ -16,6 +16,10 @@
       "desc": "By default the current date is displayed using outlined button - **show-current** prop allows you to remove the border or select different date to be displayed as the current one.",
       "heading": "### Date pickers - current date indicator"
     },
+    "date-button-events": {
+      "desc": "You can handle DOM events like `click`, `dblclick`, `mouseenter` etc that are triggered on each date/month/year button.",
```
Comment (johnleider) [inline packages/docs/src/lang/en/components/DatePickers.json:20]:
```suggestion
      "desc": "Handle events such as `@click`, `@dblclick`, `@mouseenter`, and more when interacting with *date, month, and year* buttons.",
```

File: packages/vuetify/src/components/VDatePicker/VDatePickerYears.ts [hunk 5]
```diff
@@ -5,8 +5,9 @@ import Colorable from '../../mixins/colorable'
 import Localable from '../../mixins/localable'
 
 // Utils
-import { createNativeLocaleFormatter } from './util'
```
Comment (johnleider) [inline packages/vuetify/src/components/VDatePicker/VDatePickerYears.ts:8]: https://www.notion.so/vuetify/Components-d863ad73563d43609c434f92b3e25f08#9127c5ebc666490eaa82bfe633e56dcb

File: packages/vuetify/src/components/VDatePicker/mixins/date-picker-table.ts [hunk 6]
```diff
@@ -11,6 +11,8 @@ import Themeable from '../../../mixins/themeable'
 // Utils
 import isDateAllowed from '../util/isDateAllowed'
 import mixins from '../../../util/mixins'
+import mergeData from '../../../util/mergeData'
```
Comment (johnleider) [inline packages/vuetify/src/components/VDatePicker/mixins/date-picker-table.ts:14]: https://www.notion.so/vuetify/Components-d863ad73563d43609c434f92b3e25f08#9127c5ebc666490eaa82bfe633e56dcb

File: packages/vuetify/src/components/VDatePicker/mixins/date-picker-table.ts [hunk 7]
```diff
@@ -90,13 +92,13 @@ export default mixins(
     genButtonEvents (value: string, isAllowed: boolean, mouseEventType: string) {
       if (this.disabled) return undefined
 
-      return {
-        click: () => {
-          isAllowed && !this.readonly && this.$emit('input', value)
-          this.$emit(`click:${mouseEventType}`, value)
+      return mergeData({
+        on: {
+          click: () => isAllowed && !this.readonly && this.$emit('input', value),
```
Comment (johnleider) [inline packages/vuetify/src/components/VDatePicker/mixins/date-picker-table.ts:97]: https://www.notion.so/vuetify/General-8ef7db0ab62a49f3a2f7cec7b67f3a44#d10855e35ea742158c874e5f08cf02ba

File: packages/vuetify/src/components/VDatePicker/util/index.ts [hunk 8]
```diff
@@ -1,9 +1,12 @@
+import { createItemTypeNativeListeners, createItemTypeListeners } from './eventHelpers'
```
Comment (johnleider) [inline packages/vuetify/src/components/VDatePicker/util/index.ts:1]: https://www.notion.so/vuetify/Components-d863ad73563d43609c434f92b3e25f08#9127c5ebc666490eaa82bfe633e56dcb

**Pull request merged**